parts/linux/cloud-init/artifacts/cse_config.sh

```diff
@@ -534,6 +551,66 @@ EOF
 # for DNS.
 iptables -I FORWARD -d 168.63.129.16 -p tcp --dport 80 -j DROP
 iptables -I FORWARD -d 168.63.129.16 -p tcp --dport 32526 -j DROP
+EOF
+
+    # As iptables rule will be cleaned every time the node is restarted, we need to ensure the rule is applied every time kubelet is started.
+    primaryNicIP=$(getPrimaryNicIP)
```

Comment:
I'd opt for making this a new script we bake onto the VHD directly and then invoke via kubelet pre-start with required parameters

Reply:
we can also add to `nodecustomdata.yaml` to ensure we don't create a dependency on a particular VHD version as well

Reply:
> I'd opt for making this a new script we bake onto the VHD directly and then invoke via kubelet pre-start with required parameters

upon further inspection, it looks like the script already knows how to resolve everything at runtime (I guess mainly the NIC IP from IMDS), thus we won't have to worry about passing in any required parameters when invoking it via ExecStartPre

Reply:
I also thought about making the script as custom data, but I encountered a question about this: the script know how to enable or disable IMDS restriction, but it don't know whether is should perform enable or disable operation, which is controlled by `ENABLE_IMDS_RESTRICTION` parameter. This parameter is provided to AgentBaker in `NodeBootstrapConfig` by AKS RP. If we move the script to custom data, how could we dynamically inject `ENABLE_IMDS_RESTRICTION` value?

Reply:
one more question, 

> we can also add to nodecustomdata.yaml to ensure we don't create a dependency on a particular VHD version as well

could you explain more on this? Per my understanding, current implementation will take effect in all Linux distributions after new AgentBakersvc is deployed(I assume all Linux distribution is using kubelet.service).

Reply:
> I also thought about making the script as custom data, but I encountered a question about this: the script know how to enable or disable IMDS restriction, but it don't know whether is should perform enable or disable operation, which is controlled by `ENABLE_IMDS_RESTRICTION` parameter. This parameter is provided to AgentBaker in `NodeBootstrapConfig` by AKS RP. If we move the script to custom data, how could we dynamically inject `ENABLE_IMDS_RESTRICTION` value?

we should be able to easily do this via kubelet service drop-in

Reply:
> one more question,
> 
> > we can also add to nodecustomdata.yaml to ensure we don't create a dependency on a particular VHD version as well
> 
> could you explain more on this? Per my understanding, current implementation will take effect in all Linux distributions after new AgentBakersvc is deployed(I assume all Linux distribution is using kubelet.service).

generally we try to avoid generating scripts at runtime (via CSE/customdata) and bake as much as possible onto the VHD ahead of time - however, there are some cases in the wild where older VHDs that don't have certain files already on-disk are bootstrapped with the _latest_ version of CSE/customdata, which assumes that those certain are files are on disk, which ends up causing breakage

in your case, I'd opt for putting the script onto the VHD via packer, while _also_ adding a new entry to the customdata payload for the script itself to make sure that even if the latest version of CSE/customdata is executed on a VHD that is missing the script, it will be generated on the fly.

Reply:
That makes great sense, refactored code as suggested, thanks for the explanation!